src/snapshot/mod.rs

```diff
@@ -0,0 +1,258 @@
+pub mod registration;
+
+use registration::{Delegations, MainnetRewardAddress, VotingRegistration};
+
+use chain_addr::{Discrimination, Kind};
+use jormungandr_lib::crypto::account::Identifier;
+use jormungandr_lib::interfaces::{Address, Initial, InitialUTxO, Value};
+use serde::Deserialize;
+use std::{collections::BTreeMap, iter::Iterator, num::NonZeroU64};
+
+pub const CATALYST_VOTING_PURPOSE_TAG: u64 = 0;
+
+#[derive(Deserialize, Clone, Debug)]
+pub struct RawSnapshot(Vec<VotingRegistration>);
+
+/// Contribution to a voting key for some registration
+#[derive(Clone, Debug, PartialEq)]
+pub struct KeyContribution {
+    pub reward_address: MainnetRewardAddress,
+    pub value: u64,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Snapshot {
+    // a raw public key is preferred so that we don't have to worry about discrimination when deserializing from
+    // a CIP-36 compatible encoding
+    inner: BTreeMap<Identifier, Vec<KeyContribution>>,
+    stake_threshold: Value,
+}
+
+impl Snapshot {
+    pub fn from_raw_snapshot(raw_snapshot: RawSnapshot, stake_threshold: Value) -> Self {
+        Self {
+            inner: raw_snapshot
+                .0
+                .into_iter()
+                .filter(|reg| reg.voting_power >= stake_threshold)
+                // TODO: add capability to select voting purpose for a snapshot.
+                // At the moment Catalyst is the only one in use
+                .filter(|reg| reg.voting_purpose == CATALYST_VOTING_PURPOSE_TAG)
+                .fold(BTreeMap::new(), |mut acc, reg| {
+                    let VotingRegistration {
+                        reward_address,
+                        delegations,
+                        voting_power,
+                        ..
+                    } = reg;
+
+                    match delegations {
+                        Delegations::Legacy(vk) => {
+                            acc.entry(vk).or_default().push(KeyContribution {
+                                reward_address,
+                                value: voting_power.into(),
+                            });
+                        }
+                        Delegations::New(mut vks) => {
+                            let voting_power = u64::from(voting_power);
+                            let total_weights =
+                                vks.iter().map(|(_vk, weight)| *weight as u64).sum::<u64>();
+
+                            let last = vks.pop().expect("CIP36 requires at least 1 delegation");
+                            let others_total_vp = vks
+                                .into_iter()
+                                .filter_map(|(vk, weight)| {
+                                    NonZeroU64::new((voting_power * weight as u64) / total_weights)
```

Comment:
I don't see anything that guarantees that `total_weights` is not 0 (because all the weights can be zero I think?), but I doubt that would happen anyway

Reply:
well spotted